### PR TITLE
fix: donation marked as complete even if card is denied by Stripe #924

### PR DESCRIPTION
**Pull request by mehul0810**
## Description
This PR resolves https://github.com/impress-org/give-recurring/issues/924

## Affects
- https://github.com/impress-org/givewp/blob/develop/includes/gateways/stripe/includes/class-give-stripe-customer.php#L371

## What to test
As per the acceptance criteria on the issue.

## Screenshots:
**Video Link:** https://www.loom.com/share/cae03ed23d1b40b6a3bcb8f0c2f4bcd2

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

File: includes/gateways/stripe/includes/class-give-stripe-customer.php [hunk 1]
```diff
@@ -374,7 +374,8 @@ public function create_customer() {
 				__( 'Stripe - Customer Creation Error', 'give' ),
 				$e->getMessage()
 			);
-
+			give_set_error( 'stripe_error', $e->getMessage() );
+			give_send_back_to_checkout( "?payment-mode={$payment_mode}&form_id={$post_data['post_data']['give-form-id']}" );
```
Comment (bradp) [inline includes/gateways/stripe/includes/class-give-stripe-customer.php:378]: Should we set this string to a variable and then pass it in, for ease of potential debugging with XDebug or other tools?

Comment (DevinWalker) [reply includes/gateways/stripe/includes/class-give-stripe-customer.php:378]: I think if you set the breakpoint to the line number you'll be able to debug it just as you would a variable.

File: includes/gateways/stripe/includes/class-give-stripe-customer.php [hunk 2]
```diff
@@ -374,7 +374,8 @@ public function create_customer() {
 				__( 'Stripe - Customer Creation Error', 'give' ),
```
Comment (ravinderk) [inline includes/gateways/stripe/includes/class-give-stripe-customer.php:374]: @mehul0810 I think we can use Stripe error code smartly to generate log title.

For example:

| error_code          |          Logically converted title |
|  --------------- |  ---------- |
| api_connection_error | Api connection error |
| api_error | Api error |
| authentication_error | Authentication error |
| card_error | Card error |
| idempotency_error | Idempotency error |
| invalid_request_error | Invalid request error |
| rate_limit_error | Rate limit error |


@DevinWalker what do you think about this?


ref: https://stripe.com/docs/api/errors

Comment (JasonTheAdams) [reply includes/gateways/stripe/includes/class-give-stripe-customer.php:374]: Yes, let's provide more specific errors. That no only helps the user but us as well when handling support.

Comment (mehul0810) [reply includes/gateways/stripe/includes/class-give-stripe-customer.php]: @ravinderk @JasonTheAdams I have created a separate issue for better error handling with Stripe. https://github.com/impress-org/givewp/issues/4725

**Comment by mehul0810** [comment]
@DevinWalker I have generalized the error message under the logs. Also, to update you that the pending donation will not be created if the customer is not created. So, failing a donation is not possible. I have double-checked it at my end. We can discuss about this on today's call if required.

File: includes/gateways/stripe/includes/class-give-stripe-customer.php [hunk 3]
```diff
@@ -367,26 +367,19 @@ public function create_customer() {
...
-			give_stripe_record_log(
-				__( 'Stripe - Customer Creation Error', 'give' ),
-				$e->getMessage()
-			);
-
 		} catch ( Exception $e ) {
+			// Something else happened.
 			give_record_gateway_error(
-				__( 'Stripe Error', 'give' ),
+				esc_html__( 'Stripe Error', 'give' ),
```
Comment (ravinderk) [inline includes/gateways/stripe/includes/class-give-stripe-customer.php:373]: @mehul0810 Update title to `Stripe Customer Creation Error`

Comment (ravinderk) [reply includes/gateways/stripe/includes/class-give-stripe-customer.php:373]: Resolved. As per @mehul0810 @DevinWalker verified this title.

**Pull request merged**